hydrolib/core/io/ini/serializer.py

```diff
@@ -0,0 +1,246 @@
+            datablock_indent). Defaults to 8.
+        datablock_spacing (int):
+            The number of spaces between datablock columns. Note that there might be
+            additional offset to ensure . is lined out. Defaults to 4.
+        comment_delimiter (str):
+            The character used to delimit comments.
+    """
+
+    section_indent: int = 0
+    property_indent: int = 4
+    datablock_indent: int = 8
+    datablock_spacing: int = 4
+    comment_delimiter: str = "#"
+
+    @property
+    def total_property_indent(self):
+        return self.section_indent + self.property_indent
+
+    @property
+    def total_datablock_indent(self):
+        return self.section_indent + self.datablock_indent
+
+
+@dataclass
+class _Lengths:
+    # Internal data class to store the lengths of the elements in a section.
+    max_key_length: int
+    max_value_length: int
+    max_datablock_element_length: Optional[Sequence[int]] = None
+
+    @staticmethod
+    def _calculate_max_length_datablock(
+        datablock: Optional[Sequence[Sequence[str]]],
+    ) -> Optional[Sequence[int]]:
+        if datablock is None or len(datablock) < 1:
+            return None
+
+        def folder(acc: Tuple, row: Sequence[str]) -> Tuple:
+            return tuple(max(curr_max, len(elem)) for curr_max, elem in zip(acc, row))
+
+        return reduce(folder, datablock, (0,) * len(datablock[0]))
+
+    @classmethod
+    def from_section(cls, section: Section):
+        properties = list(
+            prop for prop in section.content if isinstance(prop, Property)
+        )
+
+        if len(properties) > 0:
+            max_key_length = max(len(prop.key) for prop in properties)
+            max_value_length = max(
+                len(prop.value) if prop.value is not None else 0 for prop in properties
+            )
+        else:
+            max_key_length = 0
+            max_value_length = 0
+
+        return cls(
+            max_key_length=max_key_length,
+            max_value_length=max_value_length,
+            max_datablock_element_length=_Lengths._calculate_max_length_datablock(
+                section.datablock
+            ),
+        )
+
+
+class Serializer:
+    """Serializer serializes Document to its corresponding lines."""
+
+    def __init__(self, config: SerializerConfig):
+        """Creates a new Serializer with the provided configuration.
+
+        Args:
+            config (SerializerConfig): The configuration of this Serializer.
+        """
+        self._config = config
+
+    @staticmethod
+    def _interweave(iterable: Iterable, val: Any) -> Iterable:
+        # Interweave the provided iterable with the provided value:
+        # iterable_element, val, iterable_element, val, ...
+
+        # Note that this will interweave with val without making copies
+        # as such it is the same object being interweaved.
+        return chain.from_iterable(zip(iterable, repeat(val)))
```

Comment:
so the chain.from_iterable kinda flattens the tuples that come out of the zip?

Reply:
yup, that's the idea, and because zip takes elements till one of the iterables is exhausted, we can just repeat val indefinitely and it will stop when iterable is exhausted.